Disable scenarios that require a city start when city_size option set to 0

laeaton2001 opened the pull request:
Fixes #21667.
Updates:
Added a new flag to scenarios JSON to indicate that the scenario requires a city to start in.
Updated option text to include that a size of 0 disables some scenarios.
Updated character creation to sort disabled scenarios to bottom of list, color them dark grey & add explanation text to thier description.
Updated confirm action to do nothing for disbaled scenarios.

If someone can confirm I flagged the right scenarios before a merge it would be great.

BevapDin commented on src/newcharacter.cpp:
```diff
@@ -1887,8 +1901,11 @@ tab_direction set_scenario(WINDOW *w, player *u, points_left &points)
 
-        fold_and_print(w_description, 0, 0, TERMX - 2, c_green,
-                       sorted_scens[cur_id]->description(u->male).c_str());
+        std::string scenUnavailable = "This scenario is not available in this world due to city size settings. ";
```
Needs to be translated - add `_(  )`.

BevapDin commented on src/newcharacter.cpp:
```diff
@@ -1887,8 +1901,11 @@ tab_direction set_scenario(WINDOW *w, player *u, points_left &points)
+        std::string scenUnavailable = "This scenario is not available in this world due to city size settings. ";
+        std::string scenDesc = ( sorted_scens[cur_id]->has_flag( "CITY_START" ) && !scenario_sorter.cities_enabled ) ? 
+            (scenUnavailable + sorted_scens[cur_id]->description( u->male ).c_str()) :
+            sorted_scens[cur_id]->description( u->male ).c_str();
```
Why the call to `c_str` here? It should work fine without it. Same in the line above.

BevapDin commented on src/newcharacter.cpp:
```diff
@@ -2002,20 +2025,21 @@ tab_direction set_scenario(WINDOW *w, player *u, points_left &points)
-            points.skill_points -= sorted_scens[cur_id]->point_cost();
-
-
+            if( sorted_scens[cur_id]->has_flag( "CITY_START" ) && !scenario_sorter.cities_enabled ) {}
```
The styling is terrible, one can easily miss the `{}`. And the indentation is not needed, just `continue` when the selected scenario is not allowed:
```C++
if( confirm ) {
    if( scenario not allowed ) {
        continue;
    }
    ... // remaining original code.
```

BevapDin commented on src/newcharacter.cpp:
```diff
@@ -1752,8 +1754,18 @@ struct {
-
-        if (sort_by_points) {
+        if( !cities_enabled ) {
+            if( a->has_flag( "CITY_START" ) == b->has_flag( "CITY_START" ) ) {
```
You don't need to repeat all the comparisons here. You can return quickly when the two scenarios differ in the flag and otherwise continue with the existing comparison:
```C++
if( a->has_flag != b->has_flag ) {
    return a->has_flag < b->has_flag;
}
... // original code
```

laeaton2001 commented:
Thanks for the feedback.

Leland commented:
Quick screenshot for posterity:
<img width="1280" alt="screen shot 2017-10-22 at 10 43 23 am" src="https://user-images.githubusercontent.com/714017/31863034-d23905ac-b715-11e7-9210-7073dcb3ca0a.png">

Looks good. Would be much improved if the part of the description noting the disable could be made red, but I bet that's rather hardcoded.

laeaton2001 commented:
Turns out making it red wasn't hard.
![image](https://user-images.githubusercontent.com/32654313/31865471-afd0babc-b7b2-11e7-9d4b-e4cc61ed53a1.png)

Leland commented:
It's weird – I'm still getting #21667's behavior (endlessly building world) when doing the Evacuee start on this PR, but the evac shelter special doesn't spawn in the city. Wonder what the deal is.

https://github.com/CleverRaven/Cataclysm-DDA/blob/3f6775fe081493a0950970747ad55b7632fc7874/data/json/overmap_specials.json#L753-L761

laeaton2001 commented:
It's probably the road connection requirement of the evac shelter causing this since no cities means no roads. Hadn't thought of that.

laeaton2001 commented:
I've added the city_start flag to the scenarios I found where required a road for their overmap special. Since this included the Evacuee, I've updated the code to not default to that scenario if city_size is 0.

angoddu commented:
Somewhat off topic but what happens when city size is 0? Are there just no cities period or will you see the city label where a city would be? What kind of buildings would be left if there are no cities?

Leland commented:
@angoddu give it a shot! It's just wilderness. Few buildings, no cities or labels.

angoddu commented:
I've been playing on a size 2 space 8 world. Almost no cities but some. Then I can get some rags, pots, and nails. 0 would be interesting to see what is and isn't possible.
